pre-docs/writing_yaml_plans.md

```diff
@@ -473,3 +473,134 @@ plan mymodule::yamlplan(
 }
 ```
 
+## YAML plan constructs which cannot be translated to Puppet plans
+
+This section points out some of the quirks and limitations associated with converting a plan expressed as YAML to a Puppet plan. In some cases it is impossible to accurately translate from YAML to Puppet, in others code that is generated from the conversion is syntactically correct but not idiomatic Puppet. 
+
+### Named eval step
+
+The eval step allows snippets of Puppet code to be expressed in YAML plans. When converting a multi-line eval step to Puppet and storing the resulting value in a variable the `with` lambda is used. For example, the following YAML plan:
+```
+parameters:
+  foo:
+    type: Optional[Integer]
+    description: foo
+    default: 0
+
+steps:
+  - eval: |
+      $x = $foo + 1
+      $x * 2
+    name: eval_step
+
+return: $eval_step
+```
+Would be converted to:
+```
+plan yaml_plans::with_lambda(
+  Optional[Integer] $foo = 0
+) {
+  $eval_step = with() || {
+    $x = $foo + 1
+    $x * 2
+  }
+
+  return $eval_step
+}
+```
+If the author were writing the code as a Puppet plan they would likely not employ the use of the lambda. In this example the converted Puppet code is 'correct' but is not particularly natural or readable. 
+
+### Resource step variable interpolation
+
+When applying puppet resources in a resource step variable interpolation will behave differently between YAML and Puppet plans. In order to illustrate an important distinction consider the following yaml plan:
+```
+steps:
+  - description: apply prep
+    eval: >
+      apply_prep('localhost')
+  - target: localhost
+    description: Apply a file resource
+    resources:
+    - type: file
+      title: '/tmp/foo'
+      parameters:
+        content: $facts['os']['family']
+        ensure: present
+  - name: file_contents
+    description: Read contents of file managed with file resource
+    eval: >
+      file::read('/tmp/foo')
+      
+return: $file_contents
+```
+This plan performs apply prep on a localhost target, next it uses a Puppet `file` resource to write the OS family discovered from the Puppet `$facts` hash to a temporary file. Finally it reads the value written to the file and returns it. Running `bolt plan convert` on this plan produces the following puppet code:
+```
+plan yaml_plans::interpolation_pp() {
+  apply_prep('localhost')
+
+  $interpolation = apply('localhost') {
+    file { '/tmp/foo':
+      content => $facts['os']['family'],
+      ensure => 'present',
+    }
+  }
+  $file_contents = file::read('/tmp/foo')
+
+  return $file_contents
+}
+```
+In this case the Puppet plan produced from the Yaml plan works as expected where the Yaml version fails. The cause of the failure is that the `$facts` variable is attempted to be resolved as a plan variable instead of being evaluated as part of compiling the manifest code in an apply block.
```

Comment:
Minor wordsmithing and capitalization fix:

>This Puppet plan works as expected, whereas the YAML plan it was converted from fails. The failure stems from the `$facts` variable being resolved as a plan variable, instead of being evaluated as part of compiling the manifest code in an `apply` block.